### PR TITLE
[test] Check that the C/C++ module maps are looked for in `-resource-dir` before `-sdk`

### PR DESCRIPTION
**Pull request by finagolfin**
as changed in #74814

Also, dump the module map paths when `-Xfrontend -dump-clang-diagnostics` is passed in, both so we can check that manually and in these tests, and fix another Frontend test to be more specific now that this other output trips it up.

Currently, the `ClangImporter` silently adds most of these module maps to the build and there's no way to check that on the commandline.

@egorzhdan, the test I've been saying I'd write, please review.

@andrurogerz, I see you're building Android on Windows for TBC: has the linked pull caused any issues for you, as reported in those comments before?

### CODE REVIEW

File: test/ClangImporter/print-module-map-paths.swift
Comment (finagolfin) [inline test/ClangImporter/print-module-map-paths.swift:10]: I tried to make this linux C++ Interop test run natively on other platforms, including mocking up some C++ headers, but [it apparently requires some separate gcc installation also](https://github.com/llvm/llvm-project/blob/8d17824890f668cca654a6b821d642bd3192dc81/clang/lib/Driver/ToolChains/Gnu.cpp#L3422), and I didn't want to mock that too. Simply passing in include paths with `-Xcc` without that gcc directory, which is actually unused for includes but seemingly required, wouldn't work, no matter what I tried.

So I have this test only run on linux with `%target-os` and check that it's correctly redirecting these module maps to the system default `-sdk /` instead. Technically, it requires that the compiler is able to access the C++ headers, so it really requires the host OS to be linux, but since nobody is likely running the linux tests on other platforms, it should work the same. If wanted, I can split it off into a separate linux-only test file instead.

File: test/ClangImporter/print-module-map-paths.swift
Comment (finagolfin) [inline test/ClangImporter/print-module-map-paths.swift:31]: I was surprised this was used with libstdc++, but this `libcxxshim.modulemap` isn't in the static linux Musl SDK that ships libc++.

@al45tair, is this needed for all C++ stdlibs, and if so, why doesn't your Musl SDK ship it?

**Comment by egorzhdan** [comment]
@swift-ci please test

**Comment by finagolfin** [comment]
Updated to fix another test that was failing on the linux CI because of the new diagnostic output, and rebased.

**Comment by andrurogerz** [comment]
@finagolfin this is not an issue I would have encountered, but since I am set up for Android on Windows builds I will try to repro today and see how far I get.

**Comment by andrurogerz** [comment]
> @finagolfin this is not an issue I would have encountered, but since I am set up for Android on Windows builds I will try to repro today and see how far I get.

No luck-- I just run into other build issues (same as we're facing in our CI right now). @hyp do you have a working setup to build swift-firebase for an Android target on Windows?

**Comment by egorzhdan** [comment]
@swift-ci please test

**Comment by finagolfin** [comment]
Sigh, bash logic doesn't work on the Windows CI:
```
$ "if" "[" "windows-msvc" "==" "linux-gnu" "]"
# command stderr:
'if': command not found
error: command failed with exit status: 127
```
Looked at other tests and found a better way to make that test linux-only: have it check for two prefixes, one that works on all platforms and the other using `%target-os` for a few platforms.

Linux and Mac CI passed, this last change should get this passing on Windows too finally.

**Comment by egorzhdan** [comment]
@swift-ci please test

**Comment by finagolfin** [comment]
Passed CI, ready for review and merge.

**Review by egorzhdan — APPROVED** [review]
LGTM, let's also wait for an approval from @hyp or @compnerd.

File: test/ClangImporter/print-module-map-paths.swift
Comment (al45tair) [reply test/ClangImporter/print-module-map-paths.swift]: libc++ has its own module maps already; it shouldn't need shimming.

Comment (egorzhdan) [reply test/ClangImporter/print-module-map-paths.swift]: Ah sorry I missed this comment. These shims are used with all C++ stdlibs that we support, the lack of them can cause compiler crashes.

Comment (al45tair) [reply test/ClangImporter/print-module-map-paths.swift]: Ah, interesting. In which case we probably need libcxxshim.modulemap after all.

**Comment by finagolfin** [comment]
@compnerd, get a chance yet to try that updated module map pull, which was failing with your Android SDK before? This pull tests that and makes sure it's working.

**Comment by finagolfin** [comment]
@egorzhdan, I think you can go ahead and merge: the BC people don't seem too active on github lately and we can always revert both the underlying pull and this test if needed.

**Comment by finagolfin** [comment]
Ping @egorzhdan, been a month, don't think you're getting a response, might as well get this in, as it's easy to revert if needed.

**Comment by finagolfin** [comment]
Thanks Egor, been meaning to add those extra diagnostics for awhile, should come in handy when debugging C/C++ issues.

**Pull request merged**
